aspnetcore/blazor/security/additional-scenarios.md

```diff
@@ -23,12 +23,231 @@ This article explains how to configure server-side Blazor for additional securit
+To save tokens and other authentication properties in Blazor Web Apps, we recommend putting them into user claims, which can be accessed from anywhere in the app, including on the client (in the `.Client` project) when [passing authentication state](xref:blazor/security/index#manage-authentication-state-in-blazor-web-apps) and setting <xref:Microsoft.AspNetCore.Components.WebAssembly.Server.AuthenticationStateSerializationOptions.SerializeAllClaims%2A> to `true`.
 
-For Blazor Server, view the [7.0 version of this article section](xref:blazor/security/additional-scenarios?view=aspnetcore-7.0&preserve-view=true#pass-tokens-to-a-server-side-blazor-app).
+The following scenarios are covered:
+
+* [Passing tokens in apps that use an OIDC identity provider](#passing-tokens-in-apps-that-use-an-oidc-identity-provider)
+* [Passing tokens in apps that use Microsoft Identity Web packages/API for Entra ID](#passing-tokens-in-apps-that-use-microsoft-identity-web-packagesapi-for-entra-id)
+
+### Passing tokens in apps that use an OIDC identity provider
+
+For an app that adopts [OpenId Connect (OIDC) authentication](xref:blazor/security/blazor-web-app-oidc), the following example shows how to:
+
+* Retain the access token of a user in a claim when the user signs into the app.
+* Update the claim when the authentication cookie is updated with a new access token on principal validation.
+
+For a local demonstration of the approach, you can implement the following example in the sample app that accompanies <xref:blazor/security/blazor-web-app-oidc>.
+
+Where cookie authentication options (<xref:Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationOptions>) are configured, create an access token claim, which is named "`AccessToken`" in the following example, when the user signs into the app (<xref:Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationEvents.OnSigningIn%2A>):
+
+```csharp
+services.AddOptions<CookieAuthenticationOptions>(cookieScheme)
+    .Configure<CookieOidcRefresher>((cookieOptions, refresher) =>
+{
+    cookieOptions.Events.OnValidatePrincipal = context => 
+        refresher.ValidateOrRefreshCookieAsync(context, oidcScheme);
+
+    cookieOptions.Events.OnSigningIn = (context) =>
+    {
+        if (context.Principal?.Identity is not null && 
+            context.Principal.Identity.IsAuthenticated)
+        {
+            var accessToken = context.Properties.GetTokenValue("access_token");
+            var claimsIdentity = new ClaimsIdentity(context.Principal?.Identity,
+                [new Claim("AccessToken", accessToken ?? "No Access Token!")]);
+            context.Principal = new ClaimsPrincipal(claimsIdentity);
+            context.Properties.Items.Remove("access_token");
```

Comment:
I think the reason this isn't working is because GetTokenValue isn't directly indexing, it's [prefixing the key with ".Token."](https://github.com/dotnet/aspnetcore/blob/7ba32e7e88a3b8013863b1316b0ca595029d0a26/src/Http/Authentication.Abstractions/src/TokenExtensions.cs#L64). `.Remove(".Token.access_token")` would probably work.

However, if we wind up needing to use IHttpContextAccessor anyway, I'd rather just leave the access token in the AuthenticationProperties rather than putting it into a Claim. I know it was my suggestion, but I was only willing to do it to get rid of IHttpContextAccessor. And I'm not sure if it would have been worth it even then on second thought.

Reply:
> leave the access token in the AuthenticationProperties rather than putting it into a Claim

... but that seems to make getting the token(s) around the app more challenging. It's easy with the claims-based approach to get them from anywhere in the app, and it seems like it's easy to have the cookie refresher update them.